### PR TITLE
[Distributed Strategy] add teramoe for fleet

### PR DESCRIPTION
**Pull request by Wennie396**
### PR Category
Distributed Strategy

### PR Types
New features

### Description
add teramoe for fleet

### 是否引起精度变化
否

### CODE REVIEW

File: packages/paddlefleet_ops/src/paddlefleet_ops/__init__.py [hunk 1]
```diff
@@ -402,6 +408,19 @@ def _safe_load_ecosystem_lib(
         logger.warning(warning)
         blocked_import_messages["paddlefleet_ops.sonicmoe"] = error
 
+    if is_teramoe_available():
+        with paddle.use_compat_guard(
+            enable=True, scope={"teramoe"}, silent=True
+        ):
+            _safe_load_ecosystem_lib("teramoe", ops_dir, globals())
```
Comment (risemeup1111) [inline packages/paddlefleet_ops/src/paddlefleet_ops/__init__.py]: P1 这里把 TeraMoE 可用性仅绑定到硬件/版本条件，随后在顶层导入阶段无条件调用 `_safe_load_ecosystem_lib("teramoe", ...)`。本 PR 没有在 `pyproject.toml`、`packages/paddlefleet_ops` 或 third-party 中声明/打包该模块；因此满足 SM100+/CUDA 12.9/Python 3.12 的普通安装只要未预装 TeraMoE，导入 `paddlefleet_ops` 就会因 `_safe_load_ecosystem_lib` 的 ImportError 直接失败，即使 `using_teramoe=False`。请接入依赖/构建产物，或将导入失败降级为可查询的 blocked 状态。

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 2]
```diff
@@ -1522,7 +1609,23 @@ def _convert_routing_map_and_probs(
             weights, indices = paddle.topk(masked_probs, k=topk, axis=-1)
             return indices, weights
 
-        if self.using_sonic_moe:
+        if self.using_teramoe:
+            buffer = get_teramoe_buffer(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_layer.py:1632]: P1 `_init_expert_parallel()` 将 `expert_model_parallel_size <= 1` 规范化为 `self.moe_group = None`，但这个单卡分支在这里仍把该值传给 `teramoe.Buffer`。TeraMoE 的 Buffer 需要真实的进程组/通信器；因此启用 `using_teramoe` 的单卡配置会在首次 forward 失败，而新增测试用 mock 掩盖了这个路径。请明确拒绝 EP<=1，或为单卡提供受支持的 communicator，并覆盖真实初始化。

File: src/paddlefleet/transformer/moe/fused_a2a.py [hunk 3]
```diff
@@ -327,6 +336,28 @@ def reset_hybrid_ep_buffer():
...
+
+    Args:
+        group: Process group for communication
+        num_sms: Number of QPs per rank (typically matches total SMs)
+
+    Returns:
+        teramoe.Buffer: TeraMoE communication buffer
+    """
+    global _teramoe_buffer
+    if _teramoe_buffer is None or _teramoe_buffer.group != group:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/fused_a2a.py:350]: P1 这里以 `explicitly_destroy=True` 创建 Buffer，但当 group 变化时只是覆盖 `_teramoe_buffer`，全仓也没有对应的 `destroy()`/reset/teardown 调用。这样会遗留 TeraMoE 的 NVSHMEM 通信资源和 2 GiB + 1 GiB 缓冲，并在模型重建或进程组重新初始化时累积。请在替换前显式销毁旧 Buffer，并把 reset 接入训练/进程组生命周期。

File: src/paddlefleet/transformer/moe/moe_expert.py [hunk 4]
```diff
@@ -829,6 +829,48 @@ def sharded_state_dict(
         return super().sharded_state_dict(structured_name_prefix)
 
 
+class TeraMoEExpert(SonicMoEExpert):
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_expert.py:871]: P1 `TeraMoEExpert` 复用 SonicMoE 的布局转换，但 SonicMoE 在 `_convert_layout()` 中故意不转换 `weight2.main_grad`，因为 Sonic 的 wgrad kernel 通过 permute view 直接累加 grouped 布局。TeraMoE 的 `teramoe_autograd` 返回普通的 W_down 梯度，布局是当前的 `[E,H,I]`；在 `MixPrecisionLayer`/optimizer 路径中会与保留的 `[E,I,H]` `main_grad` 形状不符，或留下未更新的 master gradient。请为 TeraMoE 单独处理 weight2 的 main_grad 布局，并用真实 backward + optimizer step 回归验证。

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 5]
```diff
@@ -222,6 +228,24 @@ def __init__(
                     "paddlefleet_ops.sonicmoe"
                 ]
             )
+        if self.using_teramoe:
+            assert paddlefleet_ops.is_teramoe_available(), (
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_layer.py]: P1 这里和下方 `use_fused_weight` 检查都用 Python `assert` 承担运行时配置校验。生产以 `python -O` 启动时，unsupported TeraMoE、未启用 fused weight 等非法组合会继续构造对象，随后在 forward 以未导入模块、缺失 expert 或错误签名失败，错误信息也会丢失。请改为显式 `ValueError`/能力检查，并覆盖 `-O` 与非法配置。

File: src/paddlefleet/transformer/moe/moe_expert.py [hunk 6]
```diff
@@ -829,6 +829,48 @@ def sharded_state_dict(
...
+        **kwargs,
+    ):
+        self.convert_weights_to_sonic_layout()
+        # weight1: [E, 2*I, H] = TeraMoE W_gateup
+        # weight2: [E, H, I]   = TeraMoE W_down
+        # TeraMoE C++ kernel expects 2D input [num_tokens, hidden_size]
+        orig_shape = hidden_states.shape
+        if hidden_states.ndim == 3:
+            hidden_states = hidden_states.reshape([-1, orig_shape[-1]])
+        output = buffer.teramoe_autograd(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_expert.py:899]: P1 该调用没有校验输入/权重 dtype，而 TeraMoE kernel 只接受 BF16。`using_teramoe=True` 并不会自动把 `params_dtype`/`bf16` 限制为 BF16；FP16、FP32 或 FP8 配置会一直构造到这里，最终在自定义 autograd/kernel 中才失败。请在配置/构造阶段显式拒绝不支持的 dtype，并增加一个负向配置测试。

File: tests/single_card_tests/model/test_teramoe_moe_layer.py [hunk 7]
```diff
@@ -0,0 +1,533 @@
+# Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
```
Comment (risemeup1111) [inline tests/single_card_tests/model/test_teramoe_moe_layer.py:1]: P1 `ci/single_card_test.sh` 只遍历 `tests/single_card_tests` 下名字匹配 `test_*.py` 的文件，而本文件名是 `teramoe_moe_layer.py`。因此这 533 行新增测试不会被 CI 执行，也不能提供任何覆盖；请改名为 `test_teramoe_moe_layer.py`（或显式加入测试清单），并确认收集到用例。

File: tests/single_card_tests/model/test_teramoe_e2e_training.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        output = output * mean_score
+
+        return output
+
+    mock_buffer.teramoe_autograd.side_effect = realistic_teramoe_autograd
+    return mock_buffer
+
+
+@patch("paddlefleet.transformer.moe.moe_layer.get_teramoe_buffer")
+def run_training(mock_get_buffer):
```
Comment (risemeup1111) [inline tests/single_card_tests/model/test_teramoe_e2e_training.py]: P1 该文件会被 `ci/single_card_test.sh` 发现，但 `run_training` 不是以 `test_` 开头的 pytest 用例，文件末尾的 `if __name__ == "__main__"` 也不会由 pytest 执行，所以收集结果始终是 0 items，训练循环和日志检查都不会发生。请将逻辑封装到 `test_teramoe_e2e_training()`（可保留 CLI 入口）并验证 pytest 至少收集一个用例。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 规则来源 | 状态 |
| --- | --- | --- | --- | --- |
| 1 | [TeraMoE 子模块与构建](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803763934) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 2 | [EP 进程组校验](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803766785) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 3 | [Buffer 生命周期](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803768733) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | 🚧 |
| 4 | [TeraMoE 梯度布局](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803770765) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | 🚧 |
| 5 | [运行时 fused-weight 校验](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803772735) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 6 | [TeraMoE dtype 约束](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803774704) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 7 | [单卡测试发现](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803776764) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 8 | [EP=2 E2E 测试](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#discussion_r3803779193) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 9 | [PR 标题](https://github.com/PaddlePaddle/PaddleFleet/pull/1686#pullrequestreview-4960663304) | P3 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/e9f78b6210e52d542b7f92d798fe1ff467835273/.agents/skills/ai-review/references/base-rules.md) | ✅ |

跟进结论：TeraMoE 子模块、EP、dtype、测试入口、fused-weight 运行时校验和 PR 标题格式已修复；当前仍有 2 个 P1：显式销毁的通信 Buffer 未释放、TeraMoE `weight2.main_grad` 布局未处理，因此暂不能合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

File: tests/single_card_tests/model/test_teramoe_e2e_training.py (hunk 8 above)
Comment (risemeup1111) [reply tests/single_card_tests/model/test_teramoe_e2e_training.py]: pytest 收集问题已修复，但当前新增的 `@unittest.skipUnless(paddlefleet_ops.is_teramoe_available(), ...)` 会在单卡 CI 上跳过整个 E2E 类；该脚本随后仍按测试名检查 `test_teramoe_e2e_training_single_card.log`，跳过时不会生成该文件。请让 mock-buffer 的回归在单卡环境可执行，或为 skip 路径显式生成/跳过对应日志检查；否则这条 P1 仍未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/moe/moe_layer.py (hunk 2 above)
Comment (Wennie396) [reply src/paddlefleet/transformer/moe/moe_layer.py:1632]: Done

File: src/paddlefleet/transformer/moe/moe_layer.py (hunk 5 above)
Comment (Wennie396) [reply src/paddlefleet/transformer/moe/moe_layer.py]: Done

File: tests/single_card_tests/model/test_teramoe_moe_layer.py (hunk 7 above)
Comment (Wennie396) [reply tests/single_card_tests/model/test_teramoe_moe_layer.py:1]: Done

File: tests/single_card_tests/model/test_teramoe_e2e_training.py (hunk 8 above)
Comment (Wennie396) [reply tests/single_card_tests/model/test_teramoe_e2e_training.py]: Done

File: packages/paddlefleet_ops/src/paddlefleet_ops/__init__.py (hunk 1 above)
Comment (risemeup1111) [reply packages/paddlefleet_ops/src/paddlefleet_ops/__init__.py]: 依赖加载的降级逻辑和构建注册已经补上，但当前 commit 只修改了 `.gitmodules`，Git tree 中没有 `packages/paddlefleet_ops/third_party/TeraMoE` 的 gitlink（`git ls-files --stage` 也只有 `.gitmodules`）。同时 `build_wheel`/`build_editable` 无条件调用 `check_submodule_updated()`，该函数已把 `TeraMoE` 列为必需目录；所以干净 checkout 无法初始化该子模块，并会在构建 wheel 前直接退出。请提交对应的 submodule gitlink/固定 revision；此 P1 仍未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/moe/moe_layer.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_layer.py:1632]: 确认，当前在构造阶段显式拒绝 `expert_model_parallel_size <= 1`，并新增 EP=2 的真实多卡用例；此项已解决。

File: src/paddlefleet/transformer/moe/moe_layer.py (hunk 5 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_layer.py]: 可用性检查已改为显式 `ValueError`，但同一条问题中的 TeraMoE fused-weight 约束仍在 `moe_layer.py:440-443` 使用 `assert`；`python -O` 下该检查仍会消失。请一并改为显式异常，因此此项目前仅部分修复。

File: tests/single_card_tests/model/test_teramoe_moe_layer.py (hunk 7 above)
Comment (risemeup1111) [reply tests/single_card_tests/model/test_teramoe_moe_layer.py:1]: 确认，文件已重命名为 `test_teramoe_moe_layer.py`，符合单卡 CI 的 `test_*.py` 发现规则；此项已解决。

File: tests/single_card_tests/model/test_teramoe_e2e_training.py (hunk 8 above)
Comment (risemeup1111) [reply tests/single_card_tests/model/test_teramoe_e2e_training.py]: 确认，原单卡 E2E 文件已删除，当前改为 `test_teramoe_moe_layer_mp.py` 的真实 EP=2 forward/backward 用例，并已加入 `tests/test_configs.yaml` 的 2 卡任务；此项已解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
当前仍有两项未解决的 P1：TeraMoE Buffer 在进程组变化时未显式销毁旧资源，且 `weight2.main_grad` 仍沿用 SonicMoE 的布局特例，尚未证明普通 TeraMoE autograd 梯度能正确写回优化器。另有 PR 标题格式的 P3 建议。请修复两项 P1 后再合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
当前仍有两项未解决的 P1：TeraMoE Buffer 在进程组变化时未显式销毁旧资源，且 `weight2.main_grad` 仍沿用 SonicMoE 的布局特例，尚未证明普通 TeraMoE autograd 梯度能正确写回优化器。请修复后再合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>
